File: typescript/specialized/finiteLU.ts
import { LCG } from "../helpers/lcg.js";

let modulus = 2**19 - 1; // Mersenne prime for 32-bit integers
function modInverse(a: number, m: number): number {
  let m0 = m, x0 = 0, x1 = 1;
  if (m === 1) return 0;
  a = ((a % m) + m) % m;
  while (a > 1) {
    const q = Math.floor(a / m);
    [a, m] = [m, a % m];
    [x0, x1] = [x1 - q * x0, x0];
  }
  if (x1 < 0) x1 += m0;
  return x1;
}

function printMatrix(a: number[][]) {
  for (const row of a) {
    console.log(row.map(val => val.toFixed(3)).join(' '));
  }
}

function printVector(b: number[]) {
  console.log(b.map(val => val.toFixed(3)).join(' '));
}

function main() {
  console.log("TypeScript specialized finite field LU")
  const n = parseInt(process.argv[2] ?? "4", 10);
  console.log(`Matrix size: ${n}`);
  const rand = new LCG(987654321, 2**31 - 1, 16645, 1013904);
  const a: number[][] = Array.from({ length: n }, () => Array(n).fill(0));
  for (let i = 0; i < n; i++) {
    let rowSum = 0;
    for (let j = 0; j < n; j++) {
      if (i !== j) {
        const val = rand.nextInt();
        a[i][j] = val;
        rowSum += Math.abs(val);
      }
    }
    // Set diagonal to be strictly greater than rowSum
    a[i][i] = rowSum + rand.nextInt() + 1;
  }
  const b: number[] = Array.from({ length: n }, () => rand.nextInt());
  for (let i = 0; i < 10; i++) {
    const pivot: number[] = Array(n).fill(0);
    const aCopy = a.map(row => row.slice());
    const bCopy = b.slice();
    factor(aCopy, pivot);
    solve(aCopy, pivot, bCopy);
    console.log(`Iteration ${i} completed`);
  }
}

main();


// In-place LU factorization with pivoting, matching Rust style
export function factor(a: number[][], pivot: number[]): number {
  const n = a.length;
  const m = a[0].length;
  const min_mn = Math.min(m, n);
  for (let j = 0; j < min_mn; j++) {
    // Find pivot in column j and test for singularity
    let jp = j;
    let t = Math.abs(a[j][j]);
    for (let i = j + 1; i < m; i++) {
      const ab = Math.abs(a[i][j]);
      if (ab > t) {
        jp = i;
        t = ab;
      }
    }
    pivot[j] = jp;
    if (a[jp][j] === 0) {
      return 1;
    }
    // Swap rows j and jp if needed
    if (jp !== j) {
      const tmp = a[j];
      a[j] = a[jp];
      a[jp] = tmp;
    }
    // Compute elements j+1:M of jth column
    if (j < m - 1) {
      const recp = modInverse(a[j][j], modulus);
      for (let k = j + 1; k < m; k++) {
        a[k][j] = (a[k][j] * recp + modulus) % modulus;
      }
    }
    // Rank-1 update to trailing submatrix
    if (j < min_mn - 1) {
      for (let ii = j + 1; ii < m; ii++) {
        const aii_j = a[ii][j];
        for (let jj = j + 1; jj < n; jj++) {
          a[ii][jj] = ((a[ii][jj] - (aii_j * a[j][jj]) % modulus) + modulus) % modulus;
        }
      }
    }
  }
  return 0;
}

// Solve Ax = b using LU and pivot
export function solve(lu: number[][], pvt: number[], b: number[]): void {
  const m = lu.length;
  const n = lu[0].length;
  let ii = 0;
  for (let i = 0; i < m; i++) {
    const ip = pvt[i];
    let sum = b[ip];
    b[ip] = b[i];
    if (ii === 0) {
      for (let j = ii; j < i; j++) {
        sum = (sum - (lu[i][j] * b[j]) % modulus + modulus) % modulus;
      }
    } else if (sum === 0.0) {
      ii = i;
    }
    b[i] = sum;
  }
  for (let i = n - 1; i >= 0; i--) {
    let sum = b[i];
    for (let j = i + 1; j < n; j++) {
      sum = (sum - (lu[i][j] * b[j]) % modulus + modulus) % modulus;
    }
    b[i] = ((sum * modInverse(lu[i][i], modulus)) + modulus) % modulus;
  }
}

File: rust/src/specialized/FiniteLU.rs
use crate::helpers::lcg::Lcg;

static mut MODULUS: i32 = 2^31 -1;

fn mod_inverse(a: i32, m: i32) -> i32 {
    if a == 0 { panic! ("Inverse does not exist for zero"); }
    let mut m = m;
    let mut a = a;
    let (mut x0, mut x1) = (0, 1);
    let m0 = m;
    while a > 1 {
        let q = a / m;
        let t = m;
        m = a % m;
        a = t;
        let t = x0;
        x0 = x1 - q * x0;
        x1 = t;
    }
    if x1 < 0 { x1 += m0; }
    x1
}

fn print_matrix(a: &Vec<Vec<i32>>) {
    for row in a {
        for val in row {
            print!("{:3} ", val);
        }
        println!();
    }
    println!();
}

fn print_vector(b: &Vec<i32>) {
    for val in b {
        print!("{:3} ", val);
    }
    println!();
    println!();
}

pub fn factor(a: &mut Vec<Vec<i32>>, pivot: &mut Vec<usize>) -> i32 {
    let n = a.len();
    let m = a[0].len();
    let min_mn = std::cmp::min(m, n);
    let modulus = unsafe { MODULUS };

    for j in 0..min_mn {
        // Find pivot in column j and test for singularity
        let mut jp = j;
        let mut t = a[j][j].abs();
        for i in (j + 1)..m {
            let ab = a[i][j].abs();
            if ab > t {
                jp = i;
                t = ab;
            }
        }
        pivot[j] = jp;

        // If zero pivot, factorization fails
        if a[jp][j] == 0 {
            println!("Matrix is singular");
            return 1;
        }

        // Swap rows j and jp if needed
        if jp != j {
            a.swap(j, jp);
        }

        // Compute elements j+1:M of jth column
        if j < m - 1 {
            let recp = mod_inverse(a[j][j] as i32, modulus);
            for k in (j + 1)..m {
                a[k][j] = ((a[k][j] * recp)) % modulus;
            }
        }

        // Rank-1 update to trailing submatrix
        if j < min_mn - 1 {
            for ii in (j + 1)..m {
                let aii_j = a[ii][j];
                for jj in (j + 1)..n {
                    a[ii][jj] = ((a[ii][jj] - (aii_j * a[j][jj]) % modulus) + modulus) % modulus;
                }
            }
        }
    }
    0
}

pub fn solve(lu: &Vec<Vec<i32>>, pvt: &Vec<usize>, b: &mut Vec<i32>) {
    let m = lu.len();
    let n = lu[0].len();
    let modulus = unsafe { MODULUS };
    let mut ii = 0usize;
    for i in 0..m {
        let ip = pvt[i];
        let mut sum = b[ip];
        b[ip] = b[i];
        if ii == 0 {
            for j in ii..i {
            sum = ((sum - (lu[i][j] * b[j]) % modulus) + modulus) % modulus;
            }
        } else if sum == 0 {
            ii = i;
        }
        b[i] = sum;
    }

    for i in (0..n).rev() {
        let mut sum = b[i];
        for j in (i + 1)..n {
            sum = ((sum - (lu[i][j] * b[j]) % modulus) + modulus) % modulus;
        }
        b[i] = (sum * mod_inverse(lu[i][i], modulus)) % modulus;
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut n = 4;
    if args.len() > 1 {
        n = args[1].parse().unwrap_or(4);
    }
    unsafe{MODULUS= 2_i32.pow(13)-1};
    let modulus = unsafe{MODULUS};
    
    let mut rand = Lcg::new(12345, 1345, 16645, 1013904);
    let mut a: Vec<Vec<i32>> = vec![vec![0; n]; n];
    for i in 0..n {
        let mut row_sum = 0;
        for j in 0..n {
                let val = rand.next_int() % modulus;
                a[i][j] = val;
                row_sum += val.abs();
        }
        // Set diagonal to be strictly greater than row_sum
        a[i][i] = (row_sum + rand.next_int() + 1) % modulus;
    }
   
    let mut b: Vec<i32> = (0..n).map(|_| rand.next_int() % modulus).collect();
    print_matrix(&a);
    println!("Rust specialized finite field LU");
    println!("Matrix size: {}", n);
    for i in 0..10 {
        let mut pivot: Vec<usize> = vec![0; n];
        let mut a_copy = a.clone();
        let mut b_copy = b.clone();
        factor(&mut a_copy, &mut pivot);
        solve(&a_copy, &pivot, &mut b_copy);
        println!("Iteration {} completed", i);
        /* if (i == 9) {
            println!("Final solution: ");
            print_matrix(&a_copy);
            print_vector(&b_copy);
            print_vector(&b);
        }  */
    }

}
#[allow(dead_code)]
pub fn run_algorithm() {
    main();
}
